### ISSUE TITLE
webcamsOnlyForModerator is not applied in recording

### ISSUE THREAD

**Issue by foranastes** [bigbluebutton/bigbluebutton#9789]
**Describe the bug**
I've set webcamsOnlyForModerator=true in bigbluebutton.properties. It's working in live stream. But users can still see other users in recorded playback.

**Comment by pbdco** [comment]
Hi is this solved in 2.3? This is a big privacy issue in some cases of which participants are told that other people could not see them.

**Comment by foranastes** [comment]
No. I'm still looking for the solution.

**Comment by matiasilva** [comment]
I'm not sure how this would work in the recordings. You can't "authenticate" users as done in the meetings so you can't selectively show webcams for certain users. Any thoughts?

**Comment by ryprfpryr** [comment]
Hey @matiasilva thank you for getting into this issue. We are also looking forward to any solution mostly because of privacy concerns mentioned by @pbdco

**Comment by foranastes** [comment]
Is there a way to record moderators cam only? ---- On Thu, 06 Aug 2020 22:51:48 +0530  notifications@github.com  wrote ----
I'm not sure how this would work in the recordings. You can't "authenticate" users as done in the meetings so you can't selectively show webcams for certain users. Any thoughts?

**Comment by matiasilva** [comment]
I agree, it is non-ideal. Thinking about how such a thing can be implemented though, it would be difficult as you would need some form of authentication so to speak to make sure the student only sees what they are supposed to. This type of thing probably shouldn't be done by the recording player to prevent another layer of complexity but could perhaps be related to a front-end that provides authentication? Not sure..

**Comment by matiasilva** [comment]
> Is there a way to record moderators cam only? 

No, not that I can think of. The only way would be to prevent participants from sharing their cameras altogether.

**Comment by ffdixon** [comment]
The `webcamsOnlyForModerator` does restrict only moderators to see viewer's webcams.  Thinking about this more, one solution would be to add a [meta parameter](https://docs.bigbluebutton.org/dev/api.html#create) on the `create` API call, such as `meta_ProcessViewerWebacams=false`, that when the record and playback processing scripts run would cause them to not include any viewer's webcams in the recording.

**Comment by ryprfpryr** [comment]
@ffdixon 👏👏  sounds great thank you!!

Beside the nature of this issue, I was guessing if we follow the logic of your solution; **¿would it possible to restrict certain quality profiles only for moderators?**

**Reason of the request:**
The ISP's in our country are very very bad and even the mobile internet quality is very precarious / and even 3G and 4G is nonexistent in most remote or rural zones. 

Our classes have 1 teacher + 40 students, and for bandwidth reasons we are enabling **just 1 cameraProfile** (low quality) with bitrate as low as possible --80 kbps (most of students and teachers have VERY bad internet connections..) + of course _webcamsOnlyForModerator= true_.

A big part of our teachers are using last generation iPhones and similar mobile devices, and they use their mobile internet connection on 3G or 4G when available (slightly superior than their home's --**rural zone** connections).. BUT the big issue here is that iPhones or similar smartphones with big resolutions (like 1792 x 828 or similar) require a minimum of 150kbps to work smoothly. Big resolutions look very laggy (even with good internet connection), specially if there are big movements on the image (example: a teacher gesturing in front of a chalkboard).

We already tried to enable just 2 camera profiles (one on 80kbps --as "default" and another on 200kbps) but students sometimes change their camera quality --intentionally or accidentally as well. When this happens, the teacher bandwidth gets very saturated and sometimes is not even possible to them to give the class with their students webcams on (PS: they **need** to see their students).

**This is why it would be necessary to enable this higher bitrates only for moderators (teachers).** 

_This is what we were thinking as an ideal possibility:_

Adding a "moderatorsOnly" value in settings.yml like this example:

```
    cameraProfiles:
    - id: low
      name: Low quality
      default: false
      bitrate: 80
      **moderatorsOnly: false**

    - id: medium
      name: Medium quality
      default: true
      bitrate: 200
      **moderatorsOnly: true**
```

And then adding something like this on the custom parameters on API call: 
`userdata-bbb_moderatorsOnly_camera_profile=`

**--if set to _true_,  the cameraProfiles with moderatorsOnly=true would be NOT shown to participants (students).**

Thank you very much.

**Comment by matiasilva** [comment]
Hi, let's move this discussion to another issue @ryprfpryr to keep them separate. On the face of it, it is a good idea. I believe the best way to implement this would not be to restrict it to any one role, but rather to pass in a new parameter on join specifying a list of allowed camera profiles so that you can configure this per user on your front-end. Users would get a subset of the allowed list. Adds less settings and less clutter! You probably already know, but you can set a preferred profile here: https://docs.bigbluebutton.org/2.2/customize.html#kurento-parameters

**Comment by ryprfpryr** [comment]
Hi @matiasilva yes I'll move it to a new issue. I agree this could be passed as a list of allowed camera profiles as well, that would be great too. 
Yes, we were using userdata-bbb_preferred_camera_profile but we still can't stop users from selecting other profiles than what we really meant.

**Comment by foranastes** [comment]
Anyone has a quick workaround to record only moderator's cam or record only certain camera profiles?

**Comment by daminebenq** [comment]
I don't think that this is supposed to be an issue, the webcamsOnlyForModerator=true, is working perfectly.
I suggest creating an issue/suggestion to add the behavior described in the issue description.

**Comment by h0jeZvgoxFepBQ2C** [comment]
We have the problem that people can see the webcams of other guests, even when webcamsOnlyForModerator is set to true. Installed the BBB Server yesterday with the install script.. Any ideas? Worked on our old BBB Server.

**Comment by kepstin** [comment]
So, I think a decision is needed on how this could actually be implemented. If we want a recording to show the same thing as a viewer in the meeting would see, then the recording processing needs a method to find out, for a given meeting, whether this setting is active. And I think it can also be toggled via lock settings during a live meeting?

So this means that we'd need to add a way to record some live session settings and lock settings changes as events into the recording system.

(This would also need to be made configurable in bigbluebutton.yml, so it could be selected for a recording to include all webcams regardless of meeting settings)

**Comment by pedrobmarin** [comment]
There should be a recorded event called `WebcamsOnlyForModeratorEvent`.

https://github.com/bigbluebutton/bigbluebutton/blob/develop/akka-bbb-apps/src/main/scala/org/bigbluebutton/core/record/events/WebcamsOnlyForModeratorRecordEvent.scala#L25

**Comment by kepstin** [comment]
The implementation should probably be done by having the [create_webcam_edl](https://github.com/bigbluebutton/bigbluebutton/blob/9eb1d61e056fd90ad7dcb8697dead6de79843654/record-and-playback/core/lib/recordandplayback/generators/events.rb#L116) method track the status of the 'webcams only for moderator' setting along with whether any given video is by a moderator or participant, to decide whether to include a given video in the output.

**Comment by GuiLeme** [comment]
## What I intend to do:

Hey guys, so, I read lots of great ideas, and I think I will go for a mix of @kepstin's and @pedrobmarin's. 

What I am planning to do is both send the param `webcamsOnlyForModerator` as a message / event that is going to be interpreted by the `recording` component later and implement a toggle for the moderator to have the power to turn on and off this "lock setting". 

### Idea (just to know if it is possible to implement):

Other than that, I would like to propose another idea: As it was implemented on the PR #14360, we were able to just verify in the chat message whether it is from moderator because of the newly added parameter `role`, and I guess we could do the same around here. I don't know if that would work, because I don't really know the webcam videos' structure. But if it is similar to the chat's messages, in the way that they are events that has some properties including, in this case, `showVideo` for instance, I could implement this idea for sure.

**Comment by prlanzarin** [comment]
> What I am planning to do is both send the param webcamsOnlyForModerator as a message / event that is going to be interpreted by the recording component later and implement a toggle for the moderator to have the power to turn on and off this "lock setting".

Don't know if I got that right, but there's already a toggle/client lock setting for that - it's called `See other viewers webcams`. As previously mentioned by @pedrobmarin , lock changes during the meeting are already registered as recording events AFAIK; you'd only have to check whether it's also registered on meeting creation as well (in case it's used by config rather than the dynamic lock setting).

> Other than that, I would like to propose another idea: As it was implemented on the PR https://github.com/bigbluebutton/bigbluebutton/pull/14360, we were able to just verify in the chat message whether it is from moderator because of the newly added parameter role, and I guess we could do the same around here. I don't know if that would work, because I don't really know the webcam videos' structure. But if it is similar to the chat's messages, in the way that they are events that has some properties including, in this case, showVideo for instance, I could implement this idea for sure.

The problem with this approach is that, unlike chat messages, camera recording events are kind of a one-time thing. If you notify a camera share as a moderator camera as gets demoted along the way (or vice-versa), you'll get stuck with the initial role value. The way to guarantee consistency of camera role during the whole meeting would be to track user role changes and associate those with recorded camera events. So it'd be the same as the first approach, but more complex because we'd be adding an unneeded role attribute to the camera recording event.
